### PR TITLE
update json encoding/decoding

### PR DESCRIPTION
**Pull request by pedrocodacy**
Use play-json

### CODE REVIEW

File: src/main/scala/com/codacy/api/helpers/ResultHelper.scala [hunk 1]
```diff
@@ -18,13 +16,15 @@ class ResultHelper(implicit ast: JsonAst, astParser: Parser[String, JsonAst]) {
     val directory = outputDirectory.getOrElse(defaultOutputDirectory)
     directory.mkdirs()
     val randomUUID = UUID.randomUUID().toString
+    //implicit val resultReportEncoder: Encoder[ResultReport] = implicitly[Encoder[ResultReport]]
```
Comment (bmbferreira) [inline src/main/scala/com/codacy/api/helpers/ResultHelper.scala]: why is this here?

**Review by bmbferreira — CHANGES_REQUESTED** [review]

File: src/main/scala/com/codacy/api/helpers/ResultHelper.scala (hunk 1 above)
Comment (pedrocodacy) [reply src/main/scala/com/codacy/api/helpers/ResultHelper.scala]: ooops, forgot to remove it

File: project/Dependencies.scala [hunk 2]
```diff
@@ -2,16 +2,22 @@ import sbt._
...
+  val jodaConvert = "org.joda" % "joda-convert" % "1.8.1" //to fix jodaTime in scala 2.10
   val jgit = "org.eclipse.jgit" % "org.eclipse.jgit" % "3.6.2.201501210735-r"
-  val `rapture-json` = "com.propensive" %% "rapture-json" % raptureVersion withSources()
-  val `rapture-net` = "com.propensive" %% "rapture-net" % raptureVersion withSources()
+
+  val `rapture-net` = "com.propensive" %% "rapture-net" % "2.0.0-M8" withSources()
+
+  def playJson(scalaVersion: String) = {
+    "com.typesafe.play" %% "play-json" % (scalaVersion match {
+      case "2.10.6" => "2.6.10"
```
Comment (rtfpessoa) [inline project/Dependencies.scala]: Please use `2.4.3`

File: project/Dependencies.scala [hunk 3]
```diff
@@ -2,16 +2,22 @@ import sbt._
...
+      case _ => "2.7.0-M1"
```
Comment (rtfpessoa) [inline project/Dependencies.scala]: Use the latest stable not a milestone

**Review by rtfpessoa — CHANGES_REQUESTED** [review]

File: build.sbt [hunk 4]
```diff
@@ -6,7 +6,7 @@ name := """codacy-api-scala"""
 
 version := "1.0.3-SNAPSHOT"
 
-scalaVersion := "2.10.6"
+scalaVersion := "2.11.11"
 
 crossScalaVersions := Seq("2.10.6", "2.11.11", "2.12.3")
```
Comment (rtfpessoa) [inline build.sbt]: Let's use: `2.10.7`, `2.11.12`, `2.12.7`

File: build.sbt [hunk 5]
```diff
@@ -6,7 +6,7 @@ name := """codacy-api-scala"""
 
 version := "1.0.3-SNAPSHOT"
 
-scalaVersion := "2.10.6"
+scalaVersion := "2.11.11"
```
Comment (rtfpessoa) [inline build.sbt]: Lets use 2.10.7 here

File: build.sbt [hunk 6]
```diff
@@ -16,10 +16,10 @@ resolvers += "Typesafe maven repository" at "http://repo.typesafe.com/typesafe/m
 
 libraryDependencies ++= Seq(
   jodaTime,
+  jodaConvert,
   jgit,
-  `rapture-json`,
-  `rapture-json-play`,
   `rapture-net`,
```
Comment (rtfpessoa) [inline build.sbt]: Don't forget to replace with scalaj-http

File: build.sbt [hunk 7]
```diff
@@ -16,10 +16,10 @@ resolvers += "Typesafe maven repository" at "http://repo.typesafe.com/typesafe/m
 
 libraryDependencies ++= Seq(
   jodaTime,
```
Comment (rtfpessoa) [inline build.sbt]: Try to remove Joda completly

**Review by rtfpessoa — CHANGES_REQUESTED** [review]

File: project/Dependencies.scala [hunk 8]
```diff
@@ -2,16 +2,15 @@ import sbt._
 
 object Dependencies {
 
-  val raptureVersion = "2.0.0-M8"
-
   // Generic
-  val jodaTime = "joda-time" % "joda-time" % "2.7"
+  //play-functional from play-json depends on joda-time which only compiles if joda-convert is explicitly added :/
```
Comment (xplosunn) [inline project/Dependencies.scala]: Add to your note that this is Scala specific.

File: src/main/scala/com/codacy/api/client/CodacyClient.scala [hunk 9]
```diff
@@ -1,16 +1,16 @@
...
 
 class CodacyClient(apiUrl: Option[String] = None, apiToken: Option[String] = None,
-                   projectToken: Option[String] = None)
-                  (implicit astParser: Parser[String, JsonAst]) {
+                   projectToken: Option[String] = None) {
+
+  case class ErrorJson(error: String)
+  implicit val errorJsonFormat: Reads[ErrorJson] = Json.reads[ErrorJson]
+
+  case class PaginatedResult[T](next: Option[String], values: Seq[T])
```
Comment (xplosunn) [inline src/main/scala/com/codacy/api/client/CodacyClient.scala]: case classes grouped together on top is preferred.

File: src/main/scala/com/codacy/api/client/CodacyClient.scala [hunk 10]
```diff
@@ -21,96 +21,78 @@ class CodacyClient(apiUrl: Option[String] = None, apiToken: Option[String] = Non
   /*
    * Does an API request and parses the json output into a class
    */
-  def execute[T](request: Request[T])(implicit extractor: Extractor[T, Json]): RequestResponse[T] = {
+  def execute[T](request: Request[T])(implicit reads: Reads[T]): RequestResponse[T] = {
     get(request.endpoint) match {
-      case SuccessfulResponse(json) => SuccessfulResponse(json.as[T])
+      case SuccessfulResponse(json) => json.validate[T].fold(
+        errors => FailedResponse(JsonOps.handleDerivationFailure(errors)),
+        derived => SuccessfulResponse(derived)
```
Comment (xplosunn) [inline src/main/scala/com/codacy/api/client/CodacyClient.scala]: "derived" ?

File: src/main/scala/com/codacy/api/client/CodacyClient.scala [hunk 11]
```diff
@@ -21,96 +21,78 @@ class CodacyClient(apiUrl: Option[String] = None, apiToken: Option[String] = Non
...
 
-    val body = Http.parse(s"$remoteUrl/${request.endpoint}")
-      .query(request.queryParameters)
-      .httpPost(value, headers)
-      .slurp[Char]
+    val body = Http(s"$remoteUrl/${request.endpoint}")
+      .params(request.queryParameters)
+      .headers(headers)
+      .postData(value)
+      .asString.body
```
Comment (xplosunn) [inline src/main/scala/com/codacy/api/client/CodacyClient.scala:65]: No check for the returned code?

File: src/main/scala/com/codacy/api/helpers/FileHelper.scala [hunk 12]
```diff
@@ -37,18 +36,21 @@ object FileHelper {
...
   }
 
-  def readJsonFromFile[A](file: File)(implicit astParser: Parser[String, JsonAst], extractor: Extractor[A, Json]): Option[A] = {
+  def readJsonFromFile[A](file: File)(implicit reads: Reads[A]): Option[A] = {
     val source = Source.fromFile(file)(Codec.UTF8)
     val lines = try {
       source.mkString
     } finally source.close()
-    Json.parse(lines).as[Option[A]]
+    Json.parse(lines).validate[A].fold(
```
Comment (xplosunn) [inline src/main/scala/com/codacy/api/helpers/FileHelper.scala]: .toOption ?

File: src/main/scala/com/codacy/api/service/CoverageServices.scala [hunk 13]
```diff
@@ -1,30 +1,25 @@
...
-  // I declared it private so we can remove it easily when rapture fixes the issue: https://github.com/propensive/rapture/issues/238
-  private implicit lazy val ser = implicitly[Serializer[CoverageFileReport, Json]]
+class CoverageServices(client: CodacyClient) {
 
   def sendReport(commitUuid: String, language: String, coverageReport: CoverageReport, partial: Boolean = false): RequestResponse[RequestSuccess] = {
     val endpoint = s"coverage/$commitUuid/${language.toLowerCase}"
     val queryParams = Map("partial" -> partial.toString)
 
-    val value = Json.format(Json(coverageReport))
+    val value = Json.stringify(Json.toJson(coverageReport))
```
Comment (xplosunn) [inline src/main/scala/com/codacy/api/service/CoverageServices.scala]: "value" => "jsonString" ?

File: src/main/scala/com/codacy/api/util/JsonOps.scala [hunk 14]
```diff
@@ -0,0 +1,9 @@
+package com.codacy.api.util
+
+import play.api.libs.json.{JsError, JsPath, Json, JsonValidationError}
+
+object JsonOps {
+  def handleDerivationFailure(error: Seq[(JsPath, Seq[JsonValidationError])]): String = {
```
Comment (xplosunn) [inline src/main/scala/com/codacy/api/util/JsonOps.scala]: "handleDerivationFailure" ? There doesn't seem to be any derivation involved.

**Review by xplosunn — CHANGES_REQUESTED** [review]

File: src/main/scala/com/codacy/api/client/CodacyClient.scala (hunk 11 above)
Comment (pedrocodacy) [reply src/main/scala/com/codacy/api/client/CodacyClient.scala:65]: Handled by attempts of conversion of body to successful response

**Review by xplosunn — APPROVED** [review]

**Review by bmbferreira — APPROVED** [review]

**Review by rtfpessoa — APPROVED** [review]
